### PR TITLE
Library Minimal Metadata review

### LINKED ISSUE #28: [New Metadata Schema Request]: Library Minimal Metadata
### Contact Details

na.hong@yale.edu; Wenjin.J.Zheng@uth.tmc.edu

### Schema Name

Library Minimal Metadata

### Oranizational Usage

Library generation labs, sequencing cores, and archives of BICAN (BRAIN Initiative Cell Atlas Network) community.

### General Description

To harmonize gaps among existing library construction processes, the Library Minimal Metadata (LMM) was developed through the consensus of domain experts, including biomedical researchers, experimenters, data analysts, and informaticians. LMM is a metadata standard that aims to facilitate the interoperation and reuse of essential information and enhance the value of library-related resources.

### Suggested Reviewers

_No response_

### What format are you submitting your schema in?

.xlsx or .csv or .tsv

### Completed README

- [X] I have completed a README for this schema.

### PR DESCRIPTION
**Pull request by patrick-lloyd-ray**

### CODE REVIEW

**Comment by patrick-lloyd-ray** [comment]
- [x] add Kim as major contributor

**Comment by lydiang** [comment]
Hi @patrick-lloyd-ray 

I see one of the proposed field name is "2nd&3rd round barcode" - would be good to avoid characters like ampersand, slashes etc in the field name ...

**Comment by lydiang** [comment]
@patrick-lloyd-ray 

In the official CSV file, I see that there is sub_domain category (Receive Sample, Generate Library, Pool Library, Delivery Library Pool). 

These sub-domains are too broad and makes it difficult to tie the metadata to specific entities. In the original file, the was a "SubGroup" field can that be use instead or in addition to?

This will make it easier to users of this schema to pull out metadata specific to a particular entity.

For example, "antibody information"  is not obvious for the official CSV file that this is a property of a barcoded cell sample.

**Comment by patrick-lloyd-ray** [comment]
@lydiang, yes, that's no problem. Just to check my understanding: you would like the sub-domain categories to be changed to what were called the 'sub-group' fields in the csv. So, I'd be eliminating `receive sample`, `generate library`, `pool library` et al. and adding `amplified cDNA`, `barcoded cell sample` et al., right?

If that's the case, @hongna2008 could you please verify this change with your group as well?

**Comment by lydiang** [comment]
@patrick-lloyd-ray

The "subgroup" needs to be **at least need to be in addition to** to make it usable

**Comment by patrick-lloyd-ray** [comment]
@lydiang -- that makes sense to me. As long as there are no objections from the Task Force, I'm okay with adding or changing (just want to verify with them).

**Comment by hongna2008** [comment]
> @lydiang, yes, that's no problem. Just to check my understanding: you would like the sub-domain categories to be changed to what were called the 'sub-group' fields in the csv. So, I'd be eliminating `receive sample`, `generate library`, `pool library` et al. and adding `amplified cDNA`, `barcoded cell sample` et al., right?
> 
> If that's the case, @hongna2008 could you please verify this change with your group as well?

@patrick-lloyd-ray @lydiang The "SubGroup" field is only used to "generate library" in the current version. It is not applied to the other 3 subdomains. I think we need a discussion before we make any adjustment to these four basic domain categories which has been approved by multiple groups.

**Comment by lydiang** [comment]
Hi @hongna2008,

Another approach would be not to adjust your four basc domain at all but an include the "subgroup" in **addition** the formalization. These "subgroup" is already a vetted part of your metadata. 

Having them tag as part of the official metadata will make it useable to portals who are modeling entities at the subgroup level.

**Comment by lydiang** [comment]
@patrick-lloyd-ray, @hongna2008 

Kim and I were looking at your PR today, we believe that you might be using a old version of the library minimal metadata.
The finalized version is version 4.2
This is the link to the version
https://docs.google.com/spreadsheets/d/1mhMQz20t4dNx-Z17wk6FITiKPrK9b7Yu/edit?usp=sharing&ouid=117290496276111470013&rtpof=true&sd=true

Some of the difference that caught our eye are:
In the github CSV file you have fields

- 1st round barcodes
- 2nd and 3rd round barcode
- Antibody information

These no longer exists in version 4.2 

These fields are subsumed by "Enriched cell source barcode name" and "Histone modification marker"

@hongna2008 , can you coordinate with @patrick-lloyd-ray  to make sure the final version of metadata is used?

**Comment by hongna2008** [comment]
@lydiang Thank you for pointing this out. The latest updates by Kim were not synchronized here as this version was uploaded before that. I uploaded the latest version (named Library_Minimal_Metadata_Latesd_Version_final) in the released issue, @patrick-lloyd-ray sorry for the version confusion.

**Comment by patrick-lloyd-ray** [comment]
Thanks for the catch! I'll work on getting this updated and convert back to "ready for review" after I've made the changes.

**Comment by lydiang** [comment]
Pasting Kim's latest email here as reference

From: Kimberly Smith <KimberlyS@alleninstitute.org>
Sent: Monday, November 6, 2023 5:30 PM
To: Zheng, Wenjin J <Wenjin.J.Zheng@uth.tmc.edu>; Hong, Na <na.hong@yale.edu>; Patrick Ray <patrick.ray@alleninstitute.org>
Cc: Lydia Ng <LydiaN@alleninstitute.org>; Kimberly Smith <KimberlyS@alleninstitute.org>
Subject: Update to Library Minimal Metadata _231106-KS
 
Jim, Na, and Patrick –

 

With the link below I am providing a retroactive backfill of operational fields for the Library Minimal Metadata list.

[https://docs.google.com/spreadsheets/d/1toLUMfGNI6oulr5I2f7RzfJTaF3Es1YF/edit#gid=1498873717](https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdocs.google.com%2Fspreadsheets%2Fd%2F1toLUMfGNI6oulr5I2f7RzfJTaF3Es1YF%2Fedit%23gid%3D1498873717&data=05%7C01%7C%7C9c9bcc1350fb4c566c3408dbdf4d2447%7C32669cd6737f4b398bddd6951120d3fc%7C0%7C0%7C638349294662904718%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=VFjohECUS8hFlq42v3Kf52KRwmsyGLL7ufO7%2Fm4ORsc%3D&reserved=0)

(Library_Minimal_Metadata_Latesd_Version_update-231106_KS)

 

These include the addition of the following:

Barcode Cell Sample Method + Value Set (this is already in the SeqPortal)
Library Aliquot QC Result + Value Set (this is being deployed in SeqPortal this week)
R1/R2 Index Name Value Set additions (from Joe Nery)
 

All updates are highlighted in bright blue.

 

Please let me know if you have questions!

Kim

**Comment by patrick-lloyd-ray** [comment]
@rightbower and @djarecka and @puja-trivedi could you please verify the data types in this csv are correct based on your expectations?

**Comment by puja-trivedi** [comment]
>could you please verify the data types in this csv are correct based on your expectations?

@patrick-lloyd-ray - I think many of the fields that are currently "integer" should actually be "float" based off the data examples given in the green table. For example `amplified cDNA amplified quantity ng`, `amplified cDNA percent cDNA longer than 400bp`, and `Library input ng` are all set to "integer" but the examples given are "float". I can send you a more detailed list of the fields I think have the incorrect data type if that would be helpful.

**Comment by patrick-lloyd-ray** [comment]
> > could you please verify the data types in this csv are correct based on your expectations?
> 
> @patrick-lloyd-ray - I think many of the fields that are currently "integer" should actually be "float" based off the data examples given in the green table. For example `amplified cDNA amplified quantity ng`, `amplified cDNA percent cDNA longer than 400bp`, and `Library input ng` are all set to "integer" but the examples given are "float". I can send you a more detailed list of the fields I think have the incorrect data type if that would be helpful.

Thanks, it would be really helpful to know which are throwing errors. You can either send me a list or just make suggested changes directly in the PR, whichever is easiest for you.

File: docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+tissue sample label,2e4ca2fc-2d77-4d19-af45-d0fb7bbc2269,tissue name,,Identifier name for final intact piece of tissue before cell or nuclei prep.  This piece of tissue will be used in dissociation and has an ROI associated with it.,This is the same field as tissue sample label in the donor to alignment metadata standard.,tissue sample label,TRUE,,text,1,9/9/23,Receive Sample,Library Minimal Metadata,,,,
+amplified cDNA label,e2606a11-114e-472f-9e05-33f9b6fc3089,amplified cdna name,,"Name of a collection of cDNA molecules derived and amplified from an input barcoded_cell_sample.  These cDNA molecules represent the gene expression of each cell, with all cDNA molecules from a given cell retaining that cell's unique barcode from the cell barcoding step.  This is a necessary step for GEX methods but is not used for ATAC methods.
+",,,TRUE,,Text,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
+amplified cDNA amplified quantity ng,0db79d05-8612-4896-b9d3-eb1558841449,amplified quantity ng,,"Amount of cDNA produced after cDNA amplification measured in nanograms
+",,,TRUE,,Integer,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
```
Comment (puja-trivedi) [inline docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv]: amplified cDNA amplified quantity ng: expected data type -> float
data example provided in green table: 129.8939955

```suggestion
",,,TRUE,,Float,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
```

File: docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+amplified cDNA PCR cycles,3827634c-3f8f-4760-b358-86ce4b030238,cDNA amplification cycles,,Number of PCR cycles used during cDNA amplification for this cDNA.,,,TRUE,,Integer,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
+cDNA amplification process date,6cc333e7-9b98-497f-b7b1-eae904db2400,cDNA amplification date,,Date of cDNA amplification.,,,FALSE,,Date,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
+amplified cDNA RNA amplification pass-fail,bc62bdb2-7dc8-4404-bb84-ce0bbcae59e5,cDNA amplification pass-fail result,,Pass or Fail result based on qualitative assessment of cDNA yield and size.,,,TRUE,,ValueSet,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
+amplified cDNA percent cDNA longer than 400bp,8d150467-f69e-461c-b54c-bcfd22f581e5,cDNA amplification percent cDNA greater than 400bp,,QC metric to measure mRNA degradation of cDNA.  Higher % is higher quality starting material.  Over 400bp is used as a universal cutoff for intact (full length) vs degraded cDNA and is a common output from Bioanalyzer and Fragment Analyzer elecropheragrams.,,,TRUE,,Integer,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
```
Comment (puja-trivedi) [inline docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv]: amplified cDNA percent cDNA longer than 400bp: expected data type -> float
data example provided in green table: 1.046
```suggestion
amplified cDNA percent cDNA longer than 400bp,8d150467-f69e-461c-b54c-bcfd22f581e5,cDNA amplification percent cDNA greater than 400bp,,QC metric to measure mRNA degradation of cDNA.  Higher % is higher quality starting material.  Over 400bp is used as a universal cutoff for intact (full length) vs degraded cDNA and is a common output from Bioanalyzer and Fragment Analyzer elecropheragrams.,,,TRUE,,Float,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
```

File: docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+library avg size bp,f851eba9-56d1-4472-9d0c-d7f8bc33000a,,,"Average size of the library in terms of base pairs.  This is used to calculate the molarity before pooling and sequencing.
+",,,TRUE,,Integer,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
+library method,7b60d59e-fdd7-4b27-a2d4-cae9b69103a6,library chemistry method,,"Standardized nomenclature to describe the library method used.  This specifies the alignment method required for the library.  For example, 10xV3.1 (for RNASeq single assay), 10xMult-GEX (for RNASeq multiome assay), and 10xMult-ATAC (for ATACSeq multiome assay)",,,TRUE,,ValueSet,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
+Library concentration nm,90805b3f-f380-4f23-b159-e7eaa0c8f052,,,Concentration of library in terms of nM (nMol/L).  Number of molecules is needed for accurate pooling of the libraries and for generating the number of target reads/cell in sequencing.,,,FALSE,,Integer,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
```
Comment (puja-trivedi) [inline docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv]: Library concentration nm: expected data type -> float
data example provided in green table: 123.6382113

```suggestion
Library concentration nm,90805b3f-f380-4f23-b159-e7eaa0c8f052,,,Concentration of library in terms of nM (nMol/L).  Number of molecules is needed for accurate pooling of the libraries and for generating the number of target reads/cell in sequencing.,,,FALSE,,Float,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
```

File: docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+library creation date,9c2f575d-1b64-451d-894f-656861afe07a,library construction date,,Date of library construction,,,FALSE,,Date,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
+Library input ng,e4d31d97-722d-4771-a0e4-e6062190f2c1,,,Amount of cDNA going into library construction in nanograms.,,,TRUE,,Integer,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
```
Comment (puja-trivedi) [inline docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv]: Library input ng: expected data type -> float
data example provided in green table: 129.8939955
```suggestion
Library input ng,e4d31d97-722d-4771-a0e4-e6062190f2c1,,,Amount of cDNA going into library construction in nanograms.,,,TRUE,,Float,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
```

File: docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+library label,f717e254-3630-4342-be7b-4d56376e7afe,library name,,"Name of a library, which is a collection of fragmented and barcode-indexed DNA molecules for sequencing.  An index or barcode is typically introduced to enable identification of library origin to allow libraries to be pooled together for sequencing.",,,TRUE,,Text,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
+library prep pass-fail,6817ede2-7ead-402d-9dbc-131aca627c6c,library prep pass-fail result,,Pass or Fail result based on qualitative assessment of library yield and size.,,,TRUE,,ValueSet,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
+Library prep set,b124ffa9-9134-4a61-a30d-bb191b2fc7fa,,,"Library set, containing multiple library_names that were processed at the same time.",,,TRUE,,Text,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
+library quantification fmol,4c09ada7-c116-48bc-8fb1-0dcf5c4b939a,,,Amount of library generated in terms of femtomoles,,,TRUE,,Integer,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
```
Comment (puja-trivedi) [inline docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv]: library quantification fmol: expected data type -> float
data example provided in green table: 4327.337394
```suggestion
library quantification fmol,4c09ada7-c116-48bc-8fb1-0dcf5c4b939a,,,Amount of library generated in terms of femtomoles,,,TRUE,,Float,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
```

File: docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+library quantification ng,318b2d3a-dae7-4c63-bfbb-93862b92f63e,,,"Amount of library generated in terms of nanograms
+",,,FALSE,,Integer,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
```
Comment (puja-trivedi) [inline docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv]: library quantification ng: expected data type -> float
data example provided in green table: 1376.612572
```suggestion
",,,FALSE,,Float,1,9/9/23,Generate Library,Library Minimal Metadata,,,,
```

File: docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+library pool label,29e0578b-6427-4c93-b29b-bde27fbadeec,local pool name,,"Library lab's library pool name.  For some labs this may be the same as ""Libray pool tube local name"".   Other labs distinguish between the local tube label of the library pool and the library pool name provided to SeqCore for tracking.  Local Pool Name is used to communicate sequencing status between SeqCore and Library Labs.",,,TRUE,,Text,1,9/9/23,Delivery Library Pool,Library Minimal Metadata,,,,
+library pool tube avg size bp,cf1b7c96-cdc1-4eed-8e76-ac44fbd151f7,library pool avg size bp,,"Average insert size of library pool, measured in base pairs.",,,FALSE,,Integer,1,9/9/23,Delivery Library Pool,Library Minimal Metadata,,,,
+library pool fmol,af1a6f3f-aca9-4452-b86e-f3c70c3600b6,,,Amount of library pool in the tube as measured in femtamoles (fmol),,,FALSE,,Integer,1,9/9/23,Delivery Library Pool,Library Minimal Metadata,,,,
```
Comment (puja-trivedi) [inline docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv]: library pool fmol: expected data type -> float

```suggestion
library pool fmol,af1a6f3f-aca9-4452-b86e-f3c70c3600b6,,,Amount of library pool in the tube as measured in femtamoles (fmol),,,FALSE,,Float,1,9/9/23,Delivery Library Pool,Library Minimal Metadata,,,,
```

File: docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+Length of Index 2 (i5 Primer) ,a61c1d55-b880-499c-ba4a-30311fdca62f,,,"Separate field to replace the combined field ""Sequencing cycle"". Sequencing Cycle is needed for sequencing the library pool.  The sequencing cycle needed is specific to the Library Chemistry Method and is required instruction to the SeqCores.",,,TRUE,,Integer,1,9/9/23,Delivery Library Pool,Library Minimal Metadata,,,,
+Length of Read 1,f22ac08a-bb81-4524-91f0-0e1e1a032335,,,"Separate field to replace the combined field ""Sequencing cycle"". Sequencing Cycleis  needed for sequencing the library pool.  The sequencing cycle needed is specific to the Library Chemistry Method and is required instruction to the SeqCores.",,,TRUE,,Integer,1,9/9/23,Delivery Library Pool,Library Minimal Metadata,,,,
+library pool tube contents nM,32f2d02b-7300-4554-aa93-6de6e456eda7,library pool concentration (nM),,Library pool concentration as measured in nanomolarity (nMol/L),,,FALSE,,Float,1,9/9/23,Delivery Library Pool,Library Minimal Metadata,,,,
+Library pool tube volume ul,b1b923ac-c218-4db4-a3b1-45a219612567,,,Library pool volume as measured in ul,,,FALSE,,Integer,1,9/9/23,Delivery Library Pool,Library Minimal Metadata,,,,
```
Comment (puja-trivedi) [inline docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv]: Library pool tube volume ul: expected data type -> float
```suggestion
Library pool tube volume ul,b1b923ac-c218-4db4-a3b1-45a219612567,,,Library pool volume as measured in ul,,,FALSE,,Float,1,9/9/23,Delivery Library Pool,Library Minimal Metadata,,,,
```

File: docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+PhiX spike in percent,b5ab26ad-d523-406e-a85b-e77f2f4b06b5,,,"PhiX spike-in percent desired to be added to the library pool for sequencing.  PhiX is used to increase complexity of the sample being sequenced, to reduce sequencing artifacts maintain sequencing quality on the instruments.  This is an optional instruction to the SeqCore.",,,FALSE,,Integer,1,9/9/23,Delivery Library Pool,Library Minimal Metadata,,,,
```
Comment (puja-trivedi) [inline docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv]: PhiX spike in percent: expected data type -> float
**Note:** Usually percentages are set to float, but I don't know what PhiX spike is and if it is guaranteed to be an integer.

**Comment by puja-trivedi** [comment]
> Thanks, it would be really helpful to know which are throwing errors. You can either send me a list or just make suggested changes directly in the PR, whichever is easiest for you.

@patrick-lloyd-ray - I made suggestions on the fields I had issues with but I think @djarecka might have a few more suggestions she might want to add.

**Comment by patrick-lloyd-ray** [comment]
> > Thanks, it would be really helpful to know which are throwing errors. You can either send me a list or just make suggested changes directly in the PR, whichever is easiest for you.
> 
> @patrick-lloyd-ray - I made suggestions on the fields I had issues with but I think @djarecka might have a few more suggestions she might want to add.

Thank you! I'll wait until @djarecka chimes in before I merge.

**Comment by djarecka** [comment]
@patrick-lloyd-ray - Looks like @puja-trivedi already marked all types that had `integer` in the data type field and had float examples in the green table.

I have some additional issues with examples that are not in value sets, but looks like the list of values are not in this PR.

File: docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv (hunk 9 above)
Comment (patrick-lloyd-ray) [inline docs/schemas/Library-Minimal-Metadata/Library Minimal Metadata.csv]:
```suggestion
PhiX spike in percent,b5ab26ad-d523-406e-a85b-e77f2f4b06b5,,,"PhiX spike-in percent desired to be added to the library pool for sequencing.  PhiX is used to increase complexity of the sample being sequenced, to reduce sequencing artifacts maintain sequencing quality on the instruments.  This is an optional instruction to the SeqCore.",,,FALSE,,Float,1,9/9/23,Delivery Library Pool,Library Minimal Metadata,,,,

```

**Comment by djarecka** [comment]
@patrick-lloyd-ray - a few general comments/requests:

- in addition to column with examples and "in_subset", could we also have something like "SubGroup" (basically the name of the class the slot should belong). At the beginning I thought that this is under "sub domain name" column, but it seems like this is a name of the tab from the green table
- I'm wondering if "version" and "version_date" should be the name of the columns or just the version of entire document? Or this is meant to be the last change for the specific slot?
- I'm also not sure about the "size" and "measurement type" columns, what do you expect to keep there?

**Comment by patrick-lloyd-ray** [comment]
- [x] add units to sheet (@pbishwakarma)
- [ ] add size (@rightbower)
- [ ] add minimum value (@rightbower)
- [ ] add maximum value (@rightbower)
- [x] add statistical type (@patrick-lloyd-ray)
- [x] add in_subset (@puja-trivedi)
- [x] change is_required to nullable (@patrick-lloyd-ray)
- [x] add LinkML classes (@puja-trivedi)
- [x] add LinkML slots/attributes (@puja-trivedi)

**Pull request merged**